src/file_storage.cpp

```diff
@@ -1445,49 +1444,59 @@ namespace {
 			aux::file_entry const& fe = m_files[i];
 			TORRENT_ASSERT(fe.symlink_index < int(m_symlinks.size()));
 
-			std::string target = m_symlinks[fe.symlink_index];
-
-			// to avoid getting stuck in an infinite loop, we only allow traversing
-			// a symlink once
-			std::set<std::string> traversed;
+			if (!check_valid_symlink(fe, dir_map, dir_links, file_map)) {
+				// this symlink is invalid, make it point to itself
+				m_symlinks[fe.symlink_index] = internal_file_path(i);
+			}
+		}
+	}
 
-			// this is where we check every path element for existence. If it's not
-			// among the concrete paths, it may be a symlink, which is also OK
-			// note that we won't iterate through this for the last step, where the
-			// filename is included. The filename is validated after the loop
-			for (string_view branch = lsplit_path(target).first;
-				branch.size() < target.size();
-				branch = lsplit_path(target, branch.size() + 1).first)
-			{
-				// this is a concrete directory
-				if (dir_map.count(branch.to_string())) continue;
+	bool file_storage::check_valid_symlink(const aux::file_entry& fe,
```

Comment:
it's not obvious to me that this change can carry its own weight unless this can be made a free function. I can't easily tell whether this still depends on `this` or not. If it can be made a free function it can be unit tested, which would be a benefit.

Reply:
+